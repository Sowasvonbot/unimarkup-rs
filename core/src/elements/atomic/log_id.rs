//! Defines log-ids for atomic elements

use logid::log_id::{self, EventLevel};

use crate::{elements::log_id::LogIdSubGrp, log_id::LogIdMainGrp};

/// Atomic error log-ids for the element section
#[derive(Debug)]
pub enum AtomicErrLogId {
    /// Log-id denoting an invalid heading level
    InvalidHeadingLvl = log_id::get_log_id(
        LogIdMainGrp::Element as u8,
        LogIdSubGrp::Atomic as u8,
        EventLevel::Error,
        0,
    ),
}
