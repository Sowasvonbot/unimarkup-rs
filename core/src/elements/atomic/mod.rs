//! Contains atomic elements

mod heading;
mod paragraph;

pub use heading::*;
pub use paragraph::*;
