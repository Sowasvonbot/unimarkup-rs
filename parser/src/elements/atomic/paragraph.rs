use std::fmt::Debug;
use std::rc::Rc;

use unimarkup_commons::lexer::token::iterator::EndMatcher;
use unimarkup_inline::element::Inline;
use unimarkup_inline::inline_parser;

use crate::elements::blocks::Block;
use crate::BlockParser;

/// Structure of a Unimarkup paragraph element.
#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct Paragraph {
    /// The content of the paragraph.
    pub content: Vec<Inline>,
}

impl Paragraph {
    pub(crate) fn parse<'s, 'i>(mut parser: BlockParser<'s, 'i>) -> (BlockParser<'s, 'i>, Block) {
        let inlines = inline_parser::parse_inlines(
            &mut parser.iter,
            &mut parser.context.inline,
            None,
            Some(Rc::new(|matcher: &mut dyn EndMatcher| {
                matcher.consumed_is_blank_line()
            })),
        )
        .to_inlines();

        (parser, Block::Paragraph(Paragraph { content: inlines }))
    }
}

// impl From<Vec<&'_ Symbol<'_>>> for Paragraph {
//     fn from(value: Vec<&'_ Symbol<'_>>) -> Self {
//         let content = value
//             .iter()
//             .map(|&s| *s)
//             .collect::<Vec<Symbol<'_>>>()
//             .parse_inlines()
//             .collect();
//         let line_nr = value.get(0).map(|symbol| symbol.start.line).unwrap_or(0);

//         let id = crate::generate_id::generate_id(&format!(
//             "paragraph{delim}{}",
//             line_nr,
//             delim = types::ELEMENT_TYPE_DELIMITER
//         ))
//         .unwrap();

//         Paragraph {
//             id,
//             content,
//             attributes: None,
//             line_nr,
//         }
//     }
// }

// impl ElementParser for Paragraph {
//     type Token<'a> = &'a Symbol<'a>;

//     fn tokenize<'i>(input: &mut SymbolIterator<'i>) -> Option<TokenizeOutput<Self::Token<'i>>> {
//         let mut content_iter = input.nest(
//             None,
//             Some(Rc::new(|matcher: &mut dyn EndMatcher| {
//                 matcher.consumed_is_empty_line()
//             })),
//         );
//         let content = content_iter.take_to_end();
//         content_iter.update(input);

//         if content.is_empty() {
//             return None;
//         }

//         Some(TokenizeOutput { tokens: content })
//     }

//     fn parse(input: Vec<Self::Token<'_>>) -> Option<Blocks> {
//         let block = Block::Paragraph(Paragraph::from(input));

//         Some(vec![block])
//     }
// }
