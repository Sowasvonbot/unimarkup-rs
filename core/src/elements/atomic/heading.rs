use std::collections::HashMap;

use logid::capturing::{LogIdTracing, MappedLogId};
use logid::log_id::LogId;
use pest::iterators::{Pair, Pairs};
use pest::Span;
use strum_macros::*;
use unimarkup_inline::{Inline, ParseUnimarkupInlines};
use unimarkup_render::html::Html;
use unimarkup_render::render::Render;

use crate::elements::log_id::GeneralErrLogId;
use crate::elements::{inlines, Blocks};
use crate::frontend::parser::custom_pest_error;
use crate::frontend::parser::{self, Rule, UmParse};
use crate::log_id::CORE_LOG_ID_MAP;

/// Enum of possible heading levels for unimarkup headings
#[derive(Eq, PartialEq, Debug, strum_macros::Display, EnumString, Clone, Copy)]
#[strum(serialize_all = "kebab-case")]
pub enum HeadingLevel {
    /// Heading level 1, corresponds to `# ` in Unimarkup.
    #[strum(serialize = "level-1")]
    Level1 = 1, // start counting from 0

    /// Heading level 2, corresponds to `## ` in Unimarkup.
    #[strum(serialize = "level-2")]
    Level2,

    /// Heading level 3, corresponds to `### ` in Unimarkup.
    #[strum(serialize = "level-3")]
    Level3,

    /// Heading level 4, corresponds to `#### ` in Unimarkup.
    #[strum(serialize = "level-4")]
    Level4,

    /// Heading level 5, corresponds to `##### ` in Unimarkup.
    #[strum(serialize = "level-5")]
    Level5,

    /// Heading level 6, corresponds to `###### ` in Unimarkup.
    #[strum(serialize = "level-6")]
    Level6,

    /// Invalid level to denote an invalid heading syntax
    Invalid,
}

impl Default for HeadingLevel {
    fn default() -> Self {
        Self::Invalid
    }
}

impl From<HeadingLevel> for u8 {
    fn from(level: HeadingLevel) -> Self {
        match level {
            HeadingLevel::Level1 => 1,
            HeadingLevel::Level2 => 2,
            HeadingLevel::Level3 => 3,
            HeadingLevel::Level4 => 4,
            HeadingLevel::Level5 => 5,
            HeadingLevel::Level6 => 6,
            _ => 7,
        }
    }
}

impl From<&str> for HeadingLevel {
    fn from(input: &str) -> Self {
        match input {
            "1" | "#" => HeadingLevel::Level1,
            "2" | "##" => HeadingLevel::Level2,
            "3" | "###" => HeadingLevel::Level3,
            "4" | "####" => HeadingLevel::Level4,
            "5" | "#####" => HeadingLevel::Level5,
            "6" | "######" => HeadingLevel::Level6,
            _ => HeadingLevel::Invalid,
        }
    }
}

impl From<usize> for HeadingLevel {
    fn from(level_depth: usize) -> Self {
        match level_depth {
            1 => Self::Level1,
            2 => Self::Level2,
            3 => Self::Level3,
            4 => Self::Level4,
            5 => Self::Level5,
            6 => Self::Level6,
            _ => Self::Invalid,
        }
    }
}

/// Structure of a Unimarkup heading element.
#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub struct Heading {
    /// Unique identifier for a heading.
    pub id: String,

    /// Heading level.
    pub level: HeadingLevel,

    /// The content of the heading line.
    pub content: Vec<Inline>,

    /// Attributes of the heading.
    pub attributes: String,

    /// Line number, where the heading occurs in
    /// the Unimarkup document.
    pub line_nr: usize,
}

impl Heading {
    /// Parses a single instance of a heading element.
    fn parse_single(pair: &Pair<Rule>) -> Result<Self, MappedLogId> {
        let mut heading_data = pair.clone().into_inner();

        let heading_start = heading_data.next().expect("heading rule has heading_start");

        let heading_content = heading_data
            .next()
            .expect("heading rule has heading_content");

        let attributes = match heading_data.next() {
            Some(attrs_rule) => {
                let attributes: HashMap<&str, &str> = serde_json::from_str(attrs_rule.as_str())
                    .map_err(|err| {
                        (GeneralErrLogId::InvalidAttribute as LogId)
                            .set_event_with(
                                &CORE_LOG_ID_MAP,
                                &custom_pest_error(
                                    "Heading attributes are not valid JSON",
                                    attrs_rule.as_span(),
                                ),
                                file!(),
                                line!(),
                            )
                            .add_info(&format!("Cause: {}", err))
                    })?;

                Some(attributes)
            }
            None => None,
        };

        let level = heading_start.as_str().trim();
        let (line_nr, _) = heading_start.as_span().start_pos().line_col();

        let generated_id = match parser::generate_id(heading_content.as_str()) {
            Some(id) => id.to_lowercase(),
            None => format!("heading-{}-line-{}", level, line_nr),
        };

        let id = match attributes {
            Some(ref attrs) if attrs.get("id").is_some() => attrs.get("id").unwrap().to_string(),
            _ => generated_id,
        };

        Ok(Heading {
            id,
            level: level.into(),
            content: heading_content.as_str().parse_unimarkup_inlines().collect(),
            attributes: serde_json::to_string(&attributes.unwrap_or_default()).unwrap(),
            line_nr,
        })
    }
}

impl UmParse for Heading {
    fn parse(pairs: &mut Pairs<Rule>, span: Span) -> Result<Blocks, MappedLogId>
    where
        Self: Sized,
    {
        let heading_pairs = pairs
            .next()
            .expect("At least one pair available")
            .into_inner();

        let mut headings: Blocks = Vec::new();

        let (line_nr, _column_nr) = span.start_pos().line_col();

        for pair in heading_pairs {
            let mut heading = Self::parse_single(&pair)?;
            // child line number starts with 1
            // which leads to off by 1 error
            // hence minus 1
            heading.line_nr += line_nr - 1;
            headings.push(heading.into());
        }

        Ok(headings)
    }
}

impl AsRef<Self> for Heading {
    fn as_ref(&self) -> &Self {
        self
    }
}

impl Render for Heading {
    fn render_html(&self) -> Result<Html, MappedLogId> {
        let mut html = Html::default();

        let tag_level = u8::from(self.level).to_string();

        html.body.push_str("<h");
        html.body.push_str(&tag_level);
        html.body.push_str(" id='");
        html.body.push_str(&self.id);
        html.body.push_str("'>");

        inlines::push_inlines(&mut html, &self.content)?;

        html.body.push_str("</h");
        html.body.push_str(&tag_level);
        html.body.push('>');

        Ok(html)
    }
}

#[allow(non_snake_case)]
#[cfg(test)]
mod tests {
    use crate::elements::atomic::{Heading, HeadingLevel};
    use unimarkup_inline::ParseUnimarkupInlines;
    use unimarkup_render::render::Render;

    #[test]
    fn test__render_html__heading() {
        let lowest_level = HeadingLevel::Level1 as usize;
        let highest_level = HeadingLevel::Level6 as usize;

        for level in lowest_level..=highest_level {
            let heading_content = "This is a heading".parse_unimarkup_inlines().collect();
            let id = format!("heading-id-{}", level);

            let heading = Heading {
                id: String::from(&id),
                level: HeadingLevel::from(level),
                content: heading_content,
                attributes: String::default(),
                line_nr: level as usize,
            };

            let html = heading.render_html().unwrap();

            let expected = format!("<h{} id='{}'>This is a heading</h{}>", level, id, level);
            assert_eq!(html.body, expected);
        }
    }

    #[test]
    fn test__render_html__heading_with_inline() {
        let lowest_level = HeadingLevel::Level1 as usize;
        let highest_level = HeadingLevel::Level6 as usize;

        for level in lowest_level..=highest_level {
            let heading_content = "`This` *is _a_* **heading**"
                .parse_unimarkup_inlines()
                .collect();
            let id = format!("heading-id-{}", level);

            let heading = Heading {
                id: String::from(&id),
                level: HeadingLevel::from(level),
                content: heading_content,
                attributes: String::default(),
                line_nr: level as usize,
            };

            let html = heading.render_html().unwrap();

            let expected = format!(
                "<h{} id='{}'><pre><code>This</code></pre> <em>is <sub>a</sub></em> <strong>heading</strong></h{}>",
                level, id, level
            );
            assert_eq!(html.body, expected);
        }
    }
}
