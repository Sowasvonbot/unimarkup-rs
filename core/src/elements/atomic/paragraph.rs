use std::{
    collections::{HashMap, VecDeque},
    fmt::Debug,
};

use logid::{
    capturing::{LogIdTracing, MappedLogId},
    log_id::LogId,
};
use pest::iterators::Pairs;
use pest::Span;
use unimarkup_inline::{Inline, ParseUnimarkupInlines};
use unimarkup_render::{html::Html, render::Render};

use crate::elements::{inlines, log_id::GeneralErrLogId, UnimarkupBlocks};
use crate::{
    backend::ParseFromIr,
    elements::types::{self, ElementType},
    frontend::parser::{self, custom_pest_error, Rule, UmParse},
    log_id::CORE_LOG_ID_MAP,
    middleend::{AsIrLines, ContentIrLine},
};

/// Structure of a Unimarkup paragraph element.
#[derive(Debug, Default, Clone)]
pub struct Paragraph {
    /// Unique identifier for a paragraph.
    pub id: String,

    /// The content of the paragraph.
    pub content: Vec<Inline>,

    /// Attributes of the paragraph.
    pub attributes: String,

    /// Line number, where the paragraph occurs in
    /// the Unimarkup document.
    pub line_nr: usize,
}

impl UmParse for Paragraph {
    fn parse(pairs: &mut Pairs<Rule>, span: Span) -> Result<UnimarkupBlocks, MappedLogId>
    where
        Self: Sized,
    {
        let (line_nr, _column_nr) = span.start_pos().line_col();

        let mut paragraph_rules = pairs
            .next()
            .expect("paragraph must be there at this point")
            .into_inner();

        let content = paragraph_rules
            .next()
            .expect("Invalid paragraph: content expected")
            .as_str()
            .parse_unimarkup_inlines()
            .collect();

        let attributes = if let Some(attributes) = paragraph_rules.next() {
            let attr: HashMap<&str, &str> =
                serde_json::from_str(attributes.as_str()).map_err(|err| {
                    (GeneralErrLogId::InvalidAttribute as LogId)
                        .set_event_with(
                            &CORE_LOG_ID_MAP,
                            &custom_pest_error(
                                "Paragraph attributes are not valid JSON.",
                                attributes.as_span(),
                            ),
                            file!(),
                            line!(),
                        )
                        .add_info(&format!("Cause: {}", err))
                })?;

            Some(attr)
        } else {
            None
        };

        let id = match attributes {
            Some(ref attrs) if attrs.get("id").is_some() => attrs.get("id").unwrap().to_string(),
            _ => parser::generate_id(&format!(
                "paragraph{delim}{}",
                line_nr,
                delim = types::ELEMENT_TYPE_DELIMITER
            ))
            .unwrap(),
        };

        let paragraph_block = Paragraph {
            id,
            content,
            attributes: serde_json::to_string(&attributes.unwrap_or_default()).unwrap(),
            line_nr,
        };

        Ok(vec![Box::new(paragraph_block)])
    }
}

impl ParseFromIr for Paragraph {
    fn parse_from_ir(content_lines: &mut VecDeque<ContentIrLine>) -> Result<Self, MappedLogId>
    where
        Self: Sized,
    {
        if let Some(ir_line) = content_lines.pop_front() {
            let expected_type = ElementType::Paragraph.to_string();

            if ir_line.um_type != expected_type {
                return Err(
                    (GeneralErrLogId::InvalidElementType as LogId).set_event_with(
                        &CORE_LOG_ID_MAP,
                        &format!(
                            "Expected paragraph type to parse, instead got: '{}'",
                            ir_line.um_type
                        ),
                        file!(),
                        line!(),
                    ),
                );
            }

            let content = if !ir_line.text.is_empty() {
                &*ir_line.text
            } else {
                &*ir_line.fallback_text
            }
            .parse_unimarkup_inlines()
            .collect();

            let attributes = if !ir_line.attributes.is_empty() {
                ir_line.attributes
            } else {
                ir_line.fallback_attributes
            };

            let block = Paragraph {
                id: ir_line.id,
                content,
                attributes,
                line_nr: ir_line.line_nr,
            };

            Ok(block)
        } else {
            Err((GeneralErrLogId::FailedBlockCreation as LogId)
                .set_event_with(
                    &CORE_LOG_ID_MAP,
                    "Could not construct Paragraph.",
                    file!(),
                    line!(),
                )
                .add_cause("No content ir line available."))
        }
    }
}

impl Render for Paragraph {
    fn render_html(&self) -> Result<Html, MappedLogId> {
        let mut html = Html::default();

        html.body.push_str("<p");
        html.body.push_str(" id='");
        html.body.push_str(&self.id);
        html.body.push_str("'>");

        inlines::push_inlines(&mut html, &self.content)?;

        html.body.push_str("</p>");

        Ok(html)
    }
}

impl AsIrLines<ContentIrLine> for Paragraph {
    fn as_ir_lines(&self) -> Vec<ContentIrLine> {
        let line = ContentIrLine::new(
            &self.id,
            self.line_nr,
            ElementType::Paragraph.to_string(),
            &self
                .content
                .iter()
                .map(|inline| inline.as_string())
                .collect::<String>(),
            "",
            &self.attributes,
            "",
        );

        vec![line]
    }
}

#[allow(non_snake_case)]
#[cfg(test)]
mod tests {
    use std::collections::VecDeque;

    use unimarkup_inline::{Inline, ParseUnimarkupInlines};
    use unimarkup_render::render::Render;

    use crate::{backend::ParseFromIr, elements::types::ElementType, middleend::ContentIrLine};

    use super::Paragraph;

    #[test]
    fn test__render_html__paragraph() {
        let id = String::from("paragraph-id");
        let content: Vec<Inline> = "This is the content of the paragraph"
            .parse_unimarkup_inlines()
            .collect();

        let block = Paragraph {
            id: id.clone(),
            content: content.clone(),
            attributes: "{}".into(),
            line_nr: 0,
        };

        let expected_html = format!("<p id='{}'>{}</p>", id, content[0].as_string());

        assert_eq!(expected_html, block.render_html().unwrap().body);
    }

    #[test]
    fn test__parse_from_ir__paragraph() {
        let test_id = String::from("test-par-id");
        let content: Vec<Inline> = "This is an example paragraph\nwhich spans multiple lines"
            .parse_unimarkup_inlines()
            .collect();

        let mut lines: VecDeque<_> = vec![ContentIrLine {
            id: test_id.clone(),
            line_nr: 0,
            um_type: ElementType::Paragraph.to_string(),
            text: content.iter().map(|inline| inline.as_string()).collect(),
            attributes: String::from("{}"),
            ..Default::default()
        }]
        .into();

        let paragraph = Paragraph::parse_from_ir(&mut lines).unwrap();

        assert_eq!(paragraph.id, test_id);
        assert_eq!(paragraph.line_nr, 0);
        assert_eq!(paragraph.content, content);
        assert_eq!(paragraph.attributes, String::from("{}"));
    }

    #[test]
    fn test__render_html__paragraph_with_inline() {
        let id = String::from("paragraph-id");
        let content = "This is `the` *content* **of _the_ paragraph**"
            .parse_unimarkup_inlines()
            .collect();

        let block = Paragraph {
            id: id.clone(),
            content,
            attributes: "{}".into(),
            line_nr: 0,
        };

        let expected_html = format!(
                    "<p id='{}'>This is <pre><code>the</code></pre> <em>content</em> <strong>of <sub>the</sub> paragraph</strong></p>",
                    id
                );

        assert_eq!(expected_html, block.render_html().unwrap().body);
    }

    #[test]
    fn test__parse_from_ir__invalid_paragraph() {
        let mut lines = vec![].into();

        let block_res = Paragraph::parse_from_ir(&mut lines);

        assert!(block_res.is_err());

        let ir_line_bad_type = ContentIrLine {
            id: String::from("some-id"),
            line_nr: 2,
            um_type: format!("{}-more-info", ElementType::Paragraph),
            text: String::from("This is the text of this paragraph"),
            ..Default::default()
        };

        lines.push_front(ir_line_bad_type);

        let block_res = Paragraph::parse_from_ir(&mut lines);

        assert!(block_res.is_err());
    }
}
