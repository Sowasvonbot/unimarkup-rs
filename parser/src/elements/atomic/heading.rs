use std::rc::Rc;

use strum_macros::*;
use unimarkup_commons::lexer::token::iterator::{EndMatcher, PrefixMatcher};
use unimarkup_commons::lexer::token::TokenKind;
use unimarkup_commons::lexer::Itertools;
use unimarkup_inline::element::{Inline, InlineElement};
use unimarkup_inline::inline_parser;

use crate::elements::Blocks;
use crate::{elements::blocks::Block, BlockParser};
use unimarkup_commons::lexer::position::Position;

use super::log_id::AtomicError;

/// Enum of possible heading levels for unimarkup headings
#[derive(Eq, PartialEq, Debug, strum_macros::Display, EnumString, Clone, Copy)]
#[strum(serialize_all = "kebab-case")]
pub enum HeadingLevel {
    /// Heading level 1, corresponds to `# ` in Unimarkup.
    #[strum(serialize = "level-1")]
    Level1 = 1, // start counting from 0

    /// Heading level 2, corresponds to `## ` in Unimarkup.
    #[strum(serialize = "level-2")]
    Level2,

    /// Heading level 3, corresponds to `### ` in Unimarkup.
    #[strum(serialize = "level-3")]
    Level3,

    /// Heading level 4, corresponds to `#### ` in Unimarkup.
    #[strum(serialize = "level-4")]
    Level4,

    /// Heading level 5, corresponds to `##### ` in Unimarkup.
    #[strum(serialize = "level-5")]
    Level5,

    /// Heading level 6, corresponds to `###### ` in Unimarkup.
    #[strum(serialize = "level-6")]
    Level6,
}

impl From<HeadingLevel> for u8 {
    fn from(level: HeadingLevel) -> Self {
        match level {
            HeadingLevel::Level1 => 1,
            HeadingLevel::Level2 => 2,
            HeadingLevel::Level3 => 3,
            HeadingLevel::Level4 => 4,
            HeadingLevel::Level5 => 5,
            HeadingLevel::Level6 => 6,
        }
    }
}

impl TryFrom<&str> for HeadingLevel {
    type Error = AtomicError;

    fn try_from(input: &str) -> Result<Self, Self::Error> {
        let level = match input {
            "1" | "#" => HeadingLevel::Level1,
            "2" | "##" => HeadingLevel::Level2,
            "3" | "###" => HeadingLevel::Level3,
            "4" | "####" => HeadingLevel::Level4,
            "5" | "#####" => HeadingLevel::Level5,
            "6" | "######" => HeadingLevel::Level6,
            _ => return Err(AtomicError::InvalidHeadingLvl),
        };

        Ok(level)
    }
}

impl TryFrom<usize> for HeadingLevel {
    type Error = AtomicError;

    fn try_from(level_depth: usize) -> Result<Self, Self::Error> {
        let level = match level_depth {
            1 => Self::Level1,
            2 => Self::Level2,
            3 => Self::Level3,
            4 => Self::Level4,
            5 => Self::Level5,
            6 => Self::Level6,
            _ => return Err(AtomicError::InvalidHeadingLvl),
        };

        Ok(level)
    }
}

/// Structure of a Unimarkup heading element.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Heading {
    /// Unique identifier for a heading.
    pub id: String,

    /// Heading level.
    pub level: HeadingLevel,

    /// The content of the heading line.
    pub content: Vec<Inline>,

    /// Attributes of the heading.
    pub attributes: Option<String>,

    pub start: Position,
    pub end: Position,
}

impl Heading {
    pub(crate) fn parse<'s, 'i>(
        mut parser: BlockParser<'s, 'i>,
    ) -> (BlockParser<'s, 'i>, Option<Block>) {
        let hashes_opt = parser.iter.next();
        if hashes_opt.is_none() {
            return (parser, None);
        }
        let hashes = hashes_opt.expect("Ensured above to be Some here.");
        let hashes_len;

        if let TokenKind::Hash(len) = hashes.kind {
            // len == 0 is impossible, because TokenKind would not be of kind Hash in that case.
            if len > 6 {
                return (parser, None);
            }

            hashes_len = len;
        } else {
            return (parser, None);
        }

        if !parser.iter.consumed_matches(&[TokenKind::Space]) {
            return (parser, None);
        }

        let (hashes_prefix, spaces_prefix) = heading_prefix_sequences(hashes_len);
        let (child_hash_prefix, _) = heading_prefix_sequences(hashes_len + 1);

        let (iter, inline_context, parsed_inlines) = inline_parser::parse_inlines(
            parser.iter,
            parser.context.inline,
            Some(Rc::new(|matcher: &mut dyn PrefixMatcher| {
                matcher.consumed_prefix(hashes_prefix) || matcher.consumed_prefix(spaces_prefix)
            })),
            Some(Rc::new(|matcher: &mut dyn EndMatcher| {
                matcher.consumed_is_blank_line() || matcher.matches(child_hash_prefix)
            })),
        );
        parser.iter = iter;
        parser.context.inline = inline_context;

        if parsed_inlines.prefix_mismatch() {
            return (parser, None);
        }

        let content = parsed_inlines.to_inlines();
        let id = as_id(&content);

        //TODO: implement optional attribute parsing here

        let heading_end = parser
            .iter
            .prev()
            .expect("At least space after hash must be in prev if inlines was empty")
            .end;

        (
            parser,
            Some(Block::Heading(Heading {
                id,
                level: HeadingLevel::try_from(hashes_len)
                    .expect("Correct heading level ensured above."),
                content,
                attributes: None,
                start: hashes.start,
                end: heading_end,
            })),
        )
    }
}

fn as_id(content: &Vec<Inline>) -> String {
    let mut s = content.to_plain_string().to_lowercase();
    s = s.replace(char::is_whitespace, "-");
    s = s.replace(['\'', '"'], ""); // quotes removed to prevent early attribute closing
    s.replace('\\', "") // backslash removed to prevent html escapes
}

const HEADING_LVL_1_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(1), TokenKind::Space];
const HEADING_LVL_1_SPACE_PREFIX: [TokenKind; 2] = [TokenKind::Space, TokenKind::Space];
const HEADING_LVL_2_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(2), TokenKind::Space];
const HEADING_LVL_2_SPACE_PREFIX: [TokenKind; 3] =
    [TokenKind::Space, TokenKind::Space, TokenKind::Space];
const HEADING_LVL_3_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(3), TokenKind::Space];
const HEADING_LVL_3_SPACE_PREFIX: [TokenKind; 4] = [
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
];
const HEADING_LVL_4_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(4), TokenKind::Space];
const HEADING_LVL_4_SPACE_PREFIX: [TokenKind; 5] = [
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
];
const HEADING_LVL_5_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(5), TokenKind::Space];
const HEADING_LVL_5_SPACE_PREFIX: [TokenKind; 6] = [
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
];
const HEADING_LVL_6_HASH_PREFIX: [TokenKind; 2] = [TokenKind::Hash(6), TokenKind::Space];
const HEADING_LVL_6_SPACE_PREFIX: [TokenKind; 7] = [
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
    TokenKind::Space,
];

const fn heading_prefix_sequences(
    hashes_len: usize,
) -> (&'static [TokenKind], &'static [TokenKind]) {
    if hashes_len == 1 {
        (&HEADING_LVL_1_HASH_PREFIX, &HEADING_LVL_1_SPACE_PREFIX)
    } else if hashes_len == 2 {
        (&HEADING_LVL_2_HASH_PREFIX, &HEADING_LVL_2_SPACE_PREFIX)
    } else if hashes_len == 3 {
        (&HEADING_LVL_3_HASH_PREFIX, &HEADING_LVL_3_SPACE_PREFIX)
    } else if hashes_len == 4 {
        (&HEADING_LVL_4_HASH_PREFIX, &HEADING_LVL_4_SPACE_PREFIX)
    } else if hashes_len == 5 {
        (&HEADING_LVL_5_HASH_PREFIX, &HEADING_LVL_5_SPACE_PREFIX)
    } else {
        (&HEADING_LVL_6_HASH_PREFIX, &HEADING_LVL_6_SPACE_PREFIX)
    }
}

// /// HeadingToken for the [`ElementParser`]
// pub enum HeadingToken<'a> {
//     /// Level of the heading
//     Level(HeadingLevel),

//     /// Content of the heading
//     Content(Vec<&'a Symbol<'a>>),

//     /// Marks the end of the heading
//     End,
// }

// impl ElementParser for Heading {
//     type Token<'a> = self::HeadingToken<'a>;

//     fn tokenize<'i>(input: &mut SymbolIterator<'i>) -> Option<TokenizeOutput<Self::Token<'i>>> {
//         let mut heading_start: Vec<SymbolKind> = input
//             .peeking_take_while(|symbol| matches!(symbol.kind, SymbolKind::Hash))
//             .map(|s| s.kind)
//             .collect();

//         let level_depth = heading_start.len();
//         let level: HeadingLevel = HeadingLevel::try_from(level_depth).ok()?;
//         if input.by_ref().nth(level_depth)?.kind != SymbolKind::Whitespace {
//             return None;
//         }

//         heading_start.push(SymbolKind::Whitespace);

//         let sub_heading_start: Vec<SymbolKind> = std::iter::repeat(SymbolKind::Hash)
//             .take(heading_start.len())
//             .chain([SymbolKind::Whitespace])
//             .collect();
//         let heading_end = move |matcher: &mut dyn EndMatcher| {
//             matcher.consumed_is_empty_line()
//                 || matcher.matches(&[SymbolKind::Eoi])
//                 || level != HeadingLevel::Level6 && matcher.matches(&sub_heading_start)
//         };

//         let whitespace_indents: Vec<SymbolKind> = std::iter::repeat(SymbolKind::Whitespace)
//             .take(heading_start.len())
//             .collect();
//         let heading_prefix = move |matcher: &mut dyn PrefixMatcher| {
//             matcher.consumed_prefix(&heading_start) || matcher.consumed_prefix(&whitespace_indents)
//         };

//         let mut content_iter =
//             input.nest(Some(Rc::new(heading_prefix)), Some(Rc::new(heading_end)));
//         let content_symbols = content_iter.take_to_end();

//         // Line prefixes violated => invalid heading syntax
//         if !content_iter.end_reached() {
//             return None;
//         }

//         content_iter.update(input);

//         let output = TokenizeOutput {
//             tokens: vec![
//                 HeadingToken::Level(level),
//                 HeadingToken::Content(content_symbols),
//                 HeadingToken::End,
//             ],
//         };

//         Some(output)
//     }

//     fn parse(input: Vec<Self::Token<'_>>) -> Option<Blocks> {
//         let HeadingToken::Level(level) = input[0] else {
//             return None;
//         };
//         let HeadingToken::Content(ref symbols) = input[1] else {
//             return None;
//         };
//         let inline_start = symbols.get(0)?.start;

//         // TODO: Adapt inline lexer to also work with Vec<&'input Symbol>
//         let content = symbols
//             .iter()
//             .map(|&s| *s)
//             .collect::<Vec<Symbol<'_>>>()
//             .parse_inlines()
//             .collect();

//         let line_nr = inline_start.line;
//         let block = Self {
//             id: String::default(),
//             level,
//             content,
//             attributes: None,
//             line_nr,
//         };

//         Some(vec![Block::Heading(block)])
//     }
// }

// impl AsRef<Self> for Heading {
//     fn as_ref(&self) -> &Self {
//         self
//     }
// }
