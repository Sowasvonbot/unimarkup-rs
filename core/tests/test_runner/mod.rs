use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use unimarkup_core::elements::atomic::Paragraph;

use crate::run_snap_test;

use self::snap_test_runner::SnapTestRunner;

pub mod as_snapshot;
pub mod snap_test_runner;
pub mod spec_test_runner;

fn get_insta_settings() -> insta::Settings {
    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_path("../spec/snapshots/");

    settings
}
// TODO: replace with Config?
#[derive(Deserialize, Clone, Serialize)]
pub struct Preamble {
    pub title: Option<String>,
    pub author: Option<String>,
    pub lang: Option<String>,
}

#[derive(Deserialize)]
pub struct TestOutputs {
    pub html: Option<String>,
    pub um: Option<String>,
}

#[derive(Deserialize)]
pub struct Test {
    pub name: String,
    pub description: Option<String>,
    pub preamble: Option<Preamble>,
    pub input: String,

    #[serde(flatten)]
    pub outputs: TestOutputs,
}

#[derive(Deserialize)]
pub struct TestGroup {
    pub spec: String,
    pub name: String,
    pub description: Option<String>,
    pub preamble: Option<Preamble>,
    pub tests: Vec<Test>,
}

macro_rules! test_parser {
    ($block_ty:ty, $file_path:literal) => {
        let path = PathBuf::from(file!());
        let mut path: PathBuf = path.strip_prefix("core").unwrap().into();

        path.set_file_name("");
        path.push($file_path);

        let mut sub_path: PathBuf = path
            .components()
            .skip_while(|component| {
                dbg!(component);
                Path::new(component) != Path::new("markup")
            })
            .skip(1)
            .collect();

        sub_path.set_extension("");

        let input = std::fs::read_to_string(&path).unwrap();

        let test_group: TestGroup = serde_yaml::from_str(&input).unwrap();

        for test in &test_group.tests {
            let mut snap_runner =
                SnapTestRunner::with_parser::<$block_ty, &str>(&test.name, &test.input)
                    .with_info(file!());

            if let Some(ref description) = test.description {
                snap_runner = snap_runner.with_description(description);
            }

            let snap_runner =
                snap_runner.with_sub_path(sub_path.to_str().expect("Invalid sub path"));

            // TODO: preamble?

            $crate::run_snap_test!(snap_runner);
        }
    };
}

#[derive(Serialize)]
struct Context {
    preamble: Preamble,
}

#[test]
pub fn test_paragraph_parser() {
    test_parser!(Paragraph, "../spec/markup/blocks/paragraph.yml");
}
