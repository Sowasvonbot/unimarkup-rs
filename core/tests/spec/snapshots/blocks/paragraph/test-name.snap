---
source: core/tests/test_runner/mod.rs
description: "Some description about the test\n"
info: core/tests/test_runner/mod.rs
---
Paragraph(
	###
	Enclosed content
	###
)
---
With input:

###
Enclosed content
###


